feat: add advanced tab attributes to api/v2/event-types

ibex088 opened the pull request:
### Overview

1. This PR introduces the `advancedTab` attribute to the following updated endpoints:

   Event Types:
     - **[POST]** `/v2/event-types`
     - **[PATCH]** `/v2/event-types/{eventTypeId}`
     - **[GET]** `/v2/event-types/{eventTypeId}` — now includes additional attributes.

   Organizations Event Types:
     - **[POST]** `/v2/organizations/{orgId}/teams/{teamId}/event-types`
     - **[PATCH]** `/v2/organizations/{orgId}/teams/{teamId}/event-types/{eventTypeId}`
     - **[GET]** `/v2/organizations/{orgId}/teams/{teamId}/event-types/{eventTypeId}` — now includes additional attributes.

2. Adds a disabled state to the following attributes:
   - `seats`
   - `recurrence`
   - `bookingLimitsCount`
   - `bookingLimitsDuration`
   - `bookingWindow`
   - `confirmationPolicy`

   Commits related to this change:
   - [83ae1a6](https://github.com/calcom/cal.com/pull/16314/commits/83ae1a620003fa8d450e25d54a2fefe032989c72)
   - [e3938af](https://github.com/calcom/cal.com/pull/16314/commits/e3938af0f14491ca14a8818d1e043f6a9c3c2f5f)

3. Refactors the codebase to implement pipes for transforming input data and uses interceptors for transforming output data. The goal is to keep the service layer cleaner by handling transformation logic at the controller level, allowing services to interact freely without duplicating transformation processes.

   Commits related to this change:
   - [d2a802d](https://github.com/calcom/cal.com/pull/16314/commits/d2a802d059f67b725eb4782659596b315a17e32e) — for user event types
   - [4526811](https://github.com/calcom/cal.com/pull/16314/commits/4526811492adc44902ccba220705b0462de79ba8) — for team event types

4. Refactors our event-type input, output, and update classes to extend from a single base class. This simplifies the process of adding new attributes in the future.

   Commit related to this change:
   - [a885234](https://github.com/calcom/cal.com/pull/16314/commits/a885234b8e3a113022f089981d418538eada7b5e)

---

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A sizable PR without self-review might be rejected).
- [x] I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com). If N/A, write N/A here and check the checkbox.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

---

## How should this be tested?

~~1. Set `"version": "1.2.3"` in `packages/platform/libraries/package.json`.~~
~~2. Add `"@calcom/platform-libraries-1.2.3": "npm:@calcom/platform-libraries@1.2.3"` to `apps/api/v2/package.json`.~~
✅ **Published 0.0.38**
3. If you encounter a `Cannot write file` error while running `api/v2`, delete the `dist` folder in `platform/types` and try running `api/v2` again.

### Testing `api/v2/event-types`:

---

#### 1. **Layout:**

   ![Screenshot](https://github.com/user-attachments/assets/9eece85d-c63f-400f-812a-9b2d6b8b831b)
   
   - **Input for the handler:**
     ```json
     {
       "metadata": {
         "enabledLayouts": ["month_view", "column_view"],
         "defaultLayout": "month_view"
       }
     }
     ```
   - **API V2:**
     ```json
     {
       "bookerLayouts": {
         "enabledLayouts": ["column", "month", "week"],
         "defaultLayout": "month"
       }
     }
     ```

---

#### 2. **Requires Confirmation - Always:**

   ![Screenshot](https://github.com/user-attachments/assets/8cd5267f-9ed1-4682-94f6-0a80a967906f)
   
   - **Input for the handler:**
     ```json
     {
       "requiresConfirmation": true,
       "requiresConfirmationWillBlockSlot": true
     }
     ```
   - **API V2:**
     ```json
     {
       "confirmationPolicy": {
         "type": "always",
         "blockCalendarForUnconfirmedBookings": true
       }
     }
     ```

---

#### 3. **Requires Confirmation - When booked with less than:**

   ![Screenshot](https://github.com/user-attachments/assets/633cb72e-1670-40ba-9431-2c3d2fe92f85)
   
   - **Input for the handler:**
     ```json
     {
       "requiresConfirmation": true,
       "requiresConfirmationWillBlockSlot": true,
       "metadata": {
         "requiresConfirmationThreshold": {
           "time": 60,
           "unit": "minutes"
         }
       }
     }
     ```
   - **API V2:**
     ```json
     {
       "confirmationPolicy": {
         "type": "always",
         "blockCalendarForUnconfirmedBookings": true,
         "noticeThreshold": {
           "count": 60,
           "unit": "minutes"
         }
       }
     }
     ```

---

#### 4. **Requires Confirmation - Disabled:**

   ![Screenshot](https://github.com/user-attachments/assets/145c3c12-66da-44c5-9742-1b44dbf7c844)
   
   - **Input for the handler:**
     ```json
     {
       "requiresConfirmation": false,
       "requiresConfirmationWillBlockSlot": false
     }
     ```
   - **API V2:**
     ```json
     {
       "confirmationPolicy": {
         "disabled": true
       }
     }
     ```

---

#### 5. **Requires Booker Email Verification:**

   ![Screenshot](https://github.com/user-attachments/assets/017e32bc-1fc5-4fd5-ae8b-c8410ee9a093)
   
   - **Input for the handler:**
     ```json
     {
       "requiresBookerEmailVerification": true
     }
     ```
   - **API V2:**
     ```json
     {
       "requiresBookerEmailVerification": true
     }
     ```

---

#### 6. **Hide Notes in Calendar:**

   ![Screenshot](https://github.com/user-attachments/assets/67aee92a-e371-401a-b8ee-d8fb6b1d64a5)
   
   - **Input for the handler:**
     ```json
     {
       "hideCalendarNotes": true
     }
     ```
   - **API V2:**
     ```json
     {
       "hideCalendarNotes": true
     }
     ```

---

#### 7. **Offer Seats:**

   ![Screenshot](https://github.com/user-attachments/assets/0262d710-1cdd-432e-90df-84aa3d26a5fa)
   
   - **Input for the handler:**
     ```json
     {
       "seatsPerTimeSlot": 4,
       "seatsShowAttendees": true,
       "seatsShowAvailabilityCount": true
     }
     ```
   - **API V2:**
     ```json
     {
       "seats": {
         "seatsPerTimeSlot": 4,
         "showAttendeeInfo": true,
         "showAvailabilityCount": true
       }
     }
     ```

---

#### 8. **Offer Seats — Disabled:**

   ![Screenshot](https://github.com/user-attachments/assets/ad15adf3-8644-4b13-89f9-4534181fe8a1)

   - **Input for the handler:**
     ```json
     {
       "seatsPerTimeSlot": null
     }
     ```
   - **API V2:**
     ```json
     {
       "seats": {
         "disabled": true
       }
     }
     ```

---

#### 9. **Lock Timezone on Booking Page:**

   ![Screenshot](https://github.com/user-attachments/assets/f6b79328-1344-4c3b-802e-6c0021867820)
   
   - **Input for the handler:**
     ```json
     {
       "lockTimeZoneToggleOnBookingPage": true
     }
     ```
   - **API V2:**
     ```json
     {
       "lockTimeZoneToggleOnBookingPage": true
     }
     ```

---

#### 10. **Event Type Color:**

   ![Screenshot](https://github.com/user-attachments/assets/26e5832a-37e6-4216-999a-aaaff4e7cb85)
   
   - **Input for the handler:**
     ```json
     {
       "eventTypeColor": {
         "lightEventTypeColor": "#292929",
         "darkEventTypeColor": "#fafafa"
       }
     }
     ```
   - **API V2:**
     ```json
     {
       "color": {
         "lightThemeHex": "#292929",
         "darkThemeHex": "#fafafa"
       }
     }
     ```

---

#### 11. **Custom Event Name:**
   
![Screenshot 2024-09-13 at 2 59 23 PM](https://github.com/user-attachments/assets/a7f82af4-2521-4a51-b075-5e3cb49acb45)

- **Input for the handler:**
   ```json
   {
     "eventName": "{Event type title} between {Organiser} and {Scheduler}"
   }
   ```
   
- **API V2:**
   ```json
   {
     "customName": "{Event type title} between {Organiser} and {Scheduler}"
   }
   ```

   **Valid Variables** that can be used in the custom event name:
   - `{Event type title}`
   - `{Organiser}`
   - `{Scheduler}`
   - `{Location}`
   - `{Organiser first name}`
   - `{Scheduler first name}`
   - `{Scheduler last name}`
   - `{Event duration}`
   - `{LOCATION}`
   - `{HOST/ATTENDEE}`
   - `{HOST}`
   - `{ATTENDEE}`
   - `{USER}`

---

#### 12. **Destination Calendar:**
   
![Screenshot 2024-09-13 at 2 58 42 PM](https://github.com/user-attachments/assets/6fd98186-97a0-4e20-b468-25a1b84aab12)

- **Input for the handler:**
   ```json
   {
     "destinationCalendar": {
       "integration": "google_calendar",
       "externalId": "c_15s99t6jcb80fsbgkss3l63c14@group.calendar.google.com"
     }
   }
   ```
   
- **API V2:**
   ```json
   {
     "destinationCalendar": {
       "integration": "google_calendar",
       "externalId": "c_15s99t6jcb80fsbgkss3l63c14@group.calendar.google.com"
     }
   }
   ```

   **Note on Testing:**
   Testing this feature requires a connected calendar. Here's how you can test it:
   
   1. Use the example app to connect a Google Calendar to your account.
   2. Retrieve the user ID of the account (found in the platform dashboard).
   3. In Postman, force refresh the access token for the user ID.
   4. Use the refreshed access token to call the following API:
      ```
      POST http://localhost:5555/v2/calendars
      ```
      This API will return a list of `destinationCalendars` connected to the account.
   5. Choose an `externalId` and `integration` from the response and use them in the create-event type payload as shown above.

---

#### 13. **Use Event Type Destination Calendar Email:**
   
![Screenshot 2024-09-13 at 2 57 54 PM](https://github.com/user-attachments/assets/2dd030b8-31fc-4311-bea5-d7f7024a92e5)

- **Input for the handler:**
   ```json
   {
     "useEventTypeDestinationCalendarEmail": true
   }
   ```
   
- **API V2:**
   ```json
   {
     "useDestinationCalendarEmail": true
   }
   ```

   **Note on Enabling:**
   To enable this, a calendar must already be connected to your account.

ibex088 committed:
added bookerLayouts

ibex088 committed:
added requiresConfirmation

ibex088 committed:
added tests

ibex088 committed:
Merge branch 'main' into feat/api-v2-extend-event-types-advanced-tab

ibex088 committed:
undo extra changes

ibex088 committed:
undo extra change

ibex088 committed:
fixed imports

ibex088 committed:
Merge branch 'main' into feat/api-v2-extend-event-types-advanced-tab

ibex088 committed:
Merge branch 'main' into feat/api-v2-extend-event-types-advanced-tab

ibex088 committed:
added requiresBookerEmailVerification

ibex088 committed:
added hideCalendarNotes

ibex088 committed:
added lockTimeZoneToggleOnBookingPage

ibex088 committed:
added eventTypeColor

ibex088 committed:
Merge branch 'main' into feat/api-v2-extend-event-types-advanced-tab

ibex088 committed:
undo extra changes

ibex088 committed:
added seats

ibex088 committed:
added requiresConfirmationWillBlockSlot to requiresConfirmation

ibex088 committed:
refactor: Implement BaseEventType class to avoid duplication across EventType input/output classes

## D-R-Y ##

ibex088 committed:
temp-fix: will revert later

ibex088 committed:
Correct placement of users array from EventTypeOutput to TeamEventTypeOutput

ibex088 committed:
Refactor: Shifted transformation logic to controllers using interceptors and pipes for a cleaner service layer.

ibex088 committed:
small fix

ibex088 committed:
Refactor: Moved transformation logic from service to controller using interceptors and pipes

** organisations/event-types

ibex088 committed:
Merge branch 'main' into feat/api-v2-extend-event-types-advanced-tab

ibex088 committed:
feat: adde tests for `validateEventTypeInputs`

ibex088 committed:
fix: failing test

supalarry commented on packages/platform/types/event-types/event-types_2024_06_14/inputs/create-event-type.input.ts:
```diff
@@ -106,9 +112,50 @@ export class CreateEventTypeInput_2024_06_14 {
+
+  @IsOptional()
+  @Type(() => EventTypeColor_2024_06_14)
+  eventTypeColor?: EventTypeColor_2024_06_14;
```
Since we already are in create event type input, i think we can rename `eventTypeColor` to simply `color`

SomayChauhan replied:
fixed - https://github.com/calcom/cal.com/commit/c58c19e6f992f52689535722045f6420ce823a79

supalarry commented on packages/platform/types/event-types/event-types_2024_06_14/inputs/requires-confirmation.input.ts:
```diff
@@ -0,0 +1,92 @@
+    description: "The time value for the notice threshold",
+    example: 30,
+  })
+  time!: number;
```
For clarity, i would put `unit` above `time` and rename `time` to `count` or `amount`.

SomayChauhan replied:
fixed https://github.com/calcom/cal.com/commit/305de6fc68d1e1c9ea62badf3e5ae88e5fc3ec60

supalarry commented on packages/platform/types/event-types/event-types_2024_06_14/outputs/event-type.output.ts:
```diff
@@ -154,106 +150,55 @@ export class EventTypeOutput_2024_06_14 {
   @IsInt()
   @IsOptional()
-  teamId?: number | null;
+  override ownerId?: number | null;
```
why override here?

SomayChauhan replied:
https://github.com/calcom/cal.com/pull/16314/files/9926a012f4c45596734197c1d2e6a83d462bcf06#r1741920893

supalarry commented on apps/api/v2/src/modules/organizations/interceptors/output-event-types-response.interceptor.ts:
```diff
@@ -0,0 +1,33 @@
+import { switchMap } from "rxjs/operators";
+
+@Injectable()
+export class OutputEventTypesResponseInterceptor<T> implements NestInterceptor {
```
I would rename it to `OutputTeamEventTypesResponseInterceptor` to distinguish it from user event type interceptor.

supalarry commented on apps/api/v2/src/modules/organizations/controllers/event-types/organizations-event-types.controller.ts:
```diff
@@ -29,40 +33,61 @@ import {
     @Body() bodyEventType: CreateTeamEventTypeInput_2024_06_14
-  ): Promise<CreateTeamEventTypeOutput> {
+  ): Promise<EventTypeHandlerResponse> {
+    await this.inputService.validateHosts(teamId, bodyEventType.hosts);
```
Why not have a function in the input service to validate hosts, transform body and validate event type inputs and then return transformed body only that is passed to create team event type.

SomayChauhan replied:
If we could figure out how to pass `teamId` to the pipe, we could convert this into a pipe that transforms the data and then validates it.

would make the code much cleaner

see https://github.com/calcom/cal.com/pull/16314/files/9926a012f4c45596734197c1d2e6a83d462bcf06#r1743564611

supalarry commented on apps/api/v2/src/ee/event-types/event-types_2024_06_14/controllers/event-types.controller.ts:
```diff
@@ -23,32 +27,53 @@ import {
   @Post("/")
   @Permissions([EVENT_TYPE_WRITE])
   @UseGuards(ApiAuthGuard)
+  @UsePipes(CreateEventTypeTransformPipe)
```
Amazing idea with the pipe!

supalarry reviewed (commented):
Amazing work @SomayChauhan ! I have some ideas below let me know what you agree / disagree with!

ibex088 committed:
fix: layout validator

SomayChauhan commented on packages/platform/types/event-types/event-types_2024_06_14/inputs/create-event-type.input.ts:
```diff
@@ -22,27 +23,32 @@ import {
+  @IsOptional()
   @IsString()
-  slug!: string;
+  slug?: string;
```
These fields were made optional in the base class because it serves as a foundation for various derived classes with different requirements. In `CreateTeamEventTypeInput_2024_06_14`, the specific input class where these fields are essential, they are overridden to be required.

This approach allows us to enforce stricter validation rules in derived classes, as loosening restrictions in derived classes is not permissible. Since we’re also extending `UpdateEventTypeInput_2024_06_14` from this base class, where these fields need to remain optional, the base class must accommodate this by making them optional. This strategy provides flexibility across different use cases while ensuring that necessary fields are mandatory where needed.

SomayChauhan commented on packages/platform/types/event-types/event-types_2024_06_14/inputs/booker-layouts.input.ts:
```diff
@@ -0,0 +1,45 @@
+export class LayoutValidator implements ValidatorConstraintInterface {
+  validate(value: any) {
+    const validValues = Object.values(BookerLayoutsInputEnum_2024_06_14);
+    return validValues.includes(value);
```
fixed in https://github.com/calcom/cal.com/commit/92557858bd85827300e3cdfbf91389ec4f7a3c93

ibex088 committed:
renamed blockCalendarForUnconfirmedBookings to blockUnconfirmedBookingsInBooker

SomayChauhan commented on packages/platform/types/event-types/event-types_2024_06_14/inputs/requires-confirmation.input.ts:
```diff
@@ -0,0 +1,92 @@
+  noticeThreshold?: NoticeThreshold_2024_06_14;
+
+  @IsBoolean()
+  blockCalendarForUnconfirmedBookings!: boolean;
```
fixed in https://github.com/calcom/cal.com/commit/58a9581ed3bcc300b376ced5c3cc9ddeaa66d8c1

ibex088 committed:
testing changes

ibex088 committed:
added disabled state

ibex088 committed:
Merge branch 'add-disable' into feat/api-v2-extend-event-types-advanced-tab

ibex088 committed:
fix merege conflict

ibex088 committed:
renamed requiresConfirmation to confirmationPolicy

ibex088 committed:
renamed OutputEventTypesResponseInterceptor to OutputTeamEventTypesResponseInterceptor

ibex088 committed:
renamed `darkThemeColor` -> `darkThemeHex` and `lightThemeColor` -> `lightThemeHex`

SomayChauhan commented on packages/platform/types/event-types/event-types_2024_06_14/inputs/event-type-color.input.ts:
```diff
@@ -0,0 +1,26 @@
+    description: "Color used for event types in light theme",
+    example: "#292929",
+  })
+  lightThemeColor!: string;
```
fixed -https://github.com/calcom/cal.com/commit/225ebb5599d291b9f5a85b37d7e6a9a4cc76e155

ibex088 committed:
renamed `eventTypeColor` -> `color`

SomayChauhan commented on apps/api/v2/src/ee/event-types/event-types_2024_06_14/controllers/event-types.controller.ts:
```diff
@@ -92,12 +120,27 @@ export class EventTypesController_2024_06_14 {
-  ): Promise<UpdateEventTypeOutput_2024_06_14> {
-    const eventType = await this.eventTypesService.updateEventType(eventTypeId, body, user);
+  ): Promise<HandlerRespose> {
+    // TODO: determine how to pass eventTypeId to the pipe and utilize the pipe in the controller
```
todo

ibex088 committed:
Merge branch 'main' into feat/api-v2-extend-event-types-advanced-tab

ibex088 committed:
changed `requiresConfirmation` -> `confirmationPolicy`

ibex088 committed:
added tests for disabled state

ibex088 committed:
Merge branch 'main' into feat/api-v2-extend-event-types-advanced-tab

ibex088 committed:
Merge branch 'main' into feat/api-v2-extend-event-types-advanced-tab

supalarry commented on packages/platform/types/event-types/event-types_2024_06_14/inputs/create-event-type.input.ts:
```diff
@@ -13,36 +13,43 @@ import {
   @IsString()
   @DocsProperty({ example: CREATE_EVENT_TITLE_EXAMPLE })
-  title!: string;
+  title?: string;
```
> These fields were made optional in the base class because it serves as a foundation for various derived classes with different requirements. In `CreateTeamEventTypeInput_2024_06_14`, the specific input class where these fields are essential, they are overridden to be required.
> 
> This approach allows us to enforce stricter validation rules in derived classes, as loosening restrictions in derived classes is not permissible. Since we’re also extending `UpdateEventTypeInput_2024_06_14` from this base class, where these fields need to remain optional, the base class must accommodate this by making them optional. This strategy provides flexibility across different use cases while ensuring that necessary fields are mandatory where needed.

I am not sure I am a fan of this approach, because if in the the base class `BaseCreateEventTypeInput_2024_06_14` everything is optional and another class extends it, it is very hard to keep track of what is needed and what is not needed. The fact that create team event type input overrides a property is a sign in my opinion that we should not derive other classes from a class where everything is optional, because if a class extends a class we assume that whatever `BaseCreateEventTypeInput_2024_06_14` can do also `CreateTeamEventTypeInput_2024_06_14` can do, but since base class has all optional we can't be sure of that.

I think it is better to have separate classes with explicitly defined properties and not have base class with all optional properties.

SomayChauhan replied:
https://github.com/calcom/cal.com/pull/16314/commits/ca1162463e75814654a63c07b4a081ec0de36d49

supalarry reviewed (commented):
Amazing PR description and good job Somay! I have some comments below:

supalarry commented on apps/api/v2/src/ee/event-types/event-types_2024_06_14/controllers/event-types.controller.e2e-spec.ts:
```diff
@@ -230,11 +236,34 @@ describe("Event types Endpoints", () => {
+        requiresConfirmation: {
+          type: ConfirmationPolicyEnum.TIME,
+          noticeThreshold: {
+            time: 60,
```
I think need update here because it was renamed to count

ibex088 committed:
added class-validators to `disabled?: false`

ibex088 committed:
Revert: Split create and update input/output classes

ibex088 committed:
feat: Refactor responses to use DTOs with plainToClass for explicit property control

	•	Updated all endpoints to return response DTOs (e.g., CreateTeamEventTypeOutput, GetTeamEventTypeOutput) to ensure consistent Swagger documentation generation.
	•	Introduced plainToClass for transforming responses, enabling explicit control over which properties are returned in the response, following Morgan’s recommendation.
	•	Replaced generic HandlerResponse type with class-validator based DTOs for each endpoint, ensuring TypeScript enforces correct response structure.
	•	Applied strategy: "excludeAll" to limit response properties to only those explicitly defined in the DTO, preventing unintended fields from being included in the API responses.

SomayChauhan commented on apps/api/v2/src/ee/event-types/event-types_2024_06_14/controllers/event-types.controller.ts:
```diff
@@ -23,32 +27,53 @@ import {
+    @Body() body: InputEventTransformed_2024_06_14,
     @GetUser() user: UserWithProfile
-  ): Promise<CreateEventTypeOutput_2024_06_14> {
+  ): Promise<HandlerRespose> {
```
https://github.com/calcom/cal.com/pull/16314/commits/ca1162463e75814654a63c07b4a081ec0de36d49

SomayChauhan commented on packages/platform/types/event-types/event-types_2024_06_14/inputs/booking-limits-duration.input.ts:
```diff
@@ -38,8 +40,11 @@ export class BookingLimitsDuration_2024_06_14 {
   })
   year?: number;
+
+  disabled?: false;
```
https://github.com/calcom/cal.com/pull/16314/commits/90ec24394ceea39965f69724db77531b5620ea1a

SomayChauhan commented on apps/api/v2/src/ee/event-types/event-types_2024_06_14/controllers/event-types.controller.ts:
```diff
@@ -78,18 +79,23 @@ export class EventTypesController_2024_06_14 {
-      data: eventType,
+      data: plainToClass(
+        EventTypeOutput_2024_06_14,
+        this.outputEventTypesService.getResponseEventType(eventType.ownerId, eventType),
```
instead of making a separate pipe, directly calling the output service here

ibex088 committed:
added pipes to transform output data

ibex088 committed:
add clean script to platform-types package.json

SomayChauhan commented on packages/platform/types/package.json:
```diff
@@ -4,7 +4,8 @@
   "scripts": {
-    "build": "tsc --build --force tsconfig.json",
+    "build": "yarn clean && tsc --build --force tsconfig.json",
+    "clean": "rm -rf ./dist",
```
added a clean script to platform-types

ibex088 committed:
Merge branch 'main' into feat/api-v2-extend-event-types-advanced-tab

ibex088 committed:
added customName

ibex088 committed:
added destination calendar

ibex088 committed:
fixed type errors

ibex088 committed:
added useDestinationCalendarEmail

ibex088 committed:
Merge branch 'main' into feat/api-v2-extend-event-types-advanced-tab

ibex088 committed:
Merge branch 'main' into feat/api-v2-extend-event-types-advanced-tab

SomayChauhan commented on apps/api/v2/src/ee/event-types/event-types_2024_06_14/services/input-event-types.service.ts:
```diff
@@ -16,16 +19,63 @@ import {
+    private readonly calendarsService: CalendarsService
   ) {}
+
+  async transformAndValidateCreateEventTypeInput(
```
https://github.com/calcom/cal.com/pull/16314#discussion_r1741880614

SomayChauhan commented:
Update: Introduced three new fields:

  1. Custom Event Name  
  2. Destination Calendar
  3. Use Event Type Destination Calendar Email

For details on testing these fields, please refer to the pull request description.

ibex088 committed:
Merge branch 'main' into feat/api-v2-extend-event-types-advanced-tab

ibex088 committed:
refactor: api-reqest and api-response

ibex088 committed:
Merge branch 'main' into feat/api-v2-extend-event-types-advanced-tab

ibex088 committed:
Update event-type.output.ts

ibex088 committed:
Merge branch 'main' into feat/api-v2-extend-event-types-advanced-tab

ibex088 committed:
fixed errors

ibex088 committed:
Update yarn.lock

SomayChauhan commented on packages/platform/types/event-types/event-types_2024_06_14/outputs/event-type.output.ts:
```diff
@@ -209,10 +220,6 @@ class BaseEventTypeOutput_2024_06_14 {
 
-  @IsBoolean()
-  @DocsProperty()
-  requiresConfirmation!: boolean;
```
renamed to confirmationPolicy

ibex088 committed:
added some missed properties to output

supalarry commented on apps/api/v2/src/ee/event-types/event-types_2024_06_14/constants/constants.ts:
```diff
@@ -1,16 +1,45 @@
+  locations: [{ type: "integrations:daily" }],
+};
+
+// Export the event types as a single object
```
No need for this comment - it simply restates what the code already tells.

supalarry commented on apps/api/v2/src/ee/event-types/event-types_2024_06_14/event-types.repository.ts:
```diff
@@ -1,38 +1,20 @@
-  slug: string;
-  locations?: ReturnType<typeof transformLocationsApiToInternal>;
-  bookingFields?: ReturnType<typeof transformBookingFieldsApiToInternal>;
+type EventTypeWithRelations = EventType & {
```
We have this type already in `apps/api/v2/src/ee/event-types/event-types_2024_06_14/services/output-event-types.service.ts`

SomayChauhan replied:
https://github.com/calcom/cal.com/commit/85b59dbdd844ce45ffe869146dc10d3baa77415e

supalarry commented on apps/api/v2/src/ee/event-types/event-types_2024_06_14/services/event-types.service.ts:
```diff
@@ -35,22 +26,13 @@ export class EventTypesService_2024_06_14 {
-      recurrence = undefined,
-      ...bodyTransformed
-    } = this.inputEventTypesService.transformInputCreateEventType(body);
+    const { destinationCalendar: _destinationCalendar, ...rest } = body;
```
Can destination calendar be omitted from `InputEventTransformed_2024_06_14` so that we don't take it out and simply don't use? What is its use if we destructure here and do nothing with it?

SomayChauhan replied:
we don't want to pass `destinationCalendar` to `createEventType` but pass to `updateEventType`

supalarry commented on apps/api/v2/src/ee/event-types/event-types_2024_06_14/services/input-event-types.service.ts:
```diff
@@ -114,7 +227,245 @@ export class InputEventTypesService_2024_06_14 {
+    requiresConfirmation?: boolean,
+    locationsDb?: ReturnType<typeof this.transformInputLocations>,
+    requiresConfirmationDb?: boolean,
+    validateDb = false
```
If we update, why do we need information about the existing event type in database? Not saying it is not necessary i just dont understand why.

Just from checking code it is hard to understand the purpose of `validateDb` . Also, if a function like `validateSeatsInput` has optional boolean its an indicator that this function should be split in 2 .

SomayChauhan replied:
https://github.com/calcom/cal.com/commit/02b4d70831c573ccacfc8f8778e9fecf26b2e123
tried to write better cleaner functions in this commit

supalarry commented on apps/api/v2/src/ee/event-types/event-types_2024_06_14/services/output-event-types.service.ts:
```diff
@@ -87,24 +109,42 @@ export class OutputEventTypesService_2024_06_14 {
+      metadata.requiresConfirmationThreshold as NoticeThresholdTransformedSchema
+    );
+    delete metadata["bookerLayouts"];
+    delete metadata["requiresConfirmationThreshold"];
```
Why are we deleting these?

SomayChauhan replied:
the reason we delete these are becase these fields are stored in metada in our db, but we have separate fields for them in our API input, so sending these properties in the metadata is a bit redundant since we are already transforming and sending these in `bookerLayouts` and `confirmationPolicy.noticeThreshold`

![Screenshot 2024-09-20 at 7 48 58 PM](https://github.com/user-attachments/assets/20ee83f4-23eb-43bc-a08e-4d176c001599)

![Screenshot 2024-09-20 at 7 50 37 PM](https://github.com/user-attachments/assets/5814ab91-422f-4265-945a-f7a827eae2c7)

supalarry commented on apps/api/v2/src/modules/organizations/repositories/organizations-event-types.repository.ts:
```diff
@@ -6,52 +7,52 @@ import { Injectable } from "@nestjs/common";
   }
 
-  async getEventTypeById(eventTypeId: number) {
+  async getEventTypeById(eventTypeId: number): Promise<DatabaseTeamEventType | null> {
```
Why do we need to explicitly typecast prisma calls? Response should be inferred and work automatically.

SomayChauhan replied:
https://github.com/calcom/cal.com/pull/16314/commits/0612d0554290adcd285defa0b07cea18bbad2949

supalarry commented on apps/api/v2/src/ee/event-types/event-types_2024_06_14/services/input-event-types.service.ts:
```diff
@@ -114,7 +227,245 @@ export class InputEventTypesService_2024_06_14 {
+      // Validate based on both provided and database values
+
+      // 1. Check if provided locations have more than one entry
+      if (locations && locations.length > 1) {
```
Why are locations and seats connected / related ?

supalarry replied:
ok this i now think about it makes sense that seats event has to have 1 location

supalarry commented on apps/api/v2/src/modules/organizations/services/event-types/input.service.ts:
```diff
@@ -19,37 +19,90 @@ export class InputOrganizationsEventTypesService {
+    const transformedBody = await this.transformInputCreateTeamEventType(teamId, inputEventType);
+
+    await this.inputEventTypesService.validateEventTypeInputs(
+      undefined,
```
Seems odd to pass undefined to a function. `validateEventTypeInputs` has 5 parameters and all of them are optional, which makes it very hard to understand which ones should be passed and how do their relate to each other any why.

If possible need to have multiple smaller functions that have all parameters required so that we can understand how do they interrelate.

SomayChauhan replied:
https://github.com/calcom/cal.com/commit/02b4d70831c573ccacfc8f8778e9fecf26b2e123

ibex088 committed:
Update constants.ts

ibex088 committed:
removed return type

ibex088 committed:
fixed some type errors

ibex088 committed:
reuse types

ibex088 committed:
Improve readability of validation functions

ibex088 committed:
Update CHANGELOG.md

ibex088 committed:
chore: post publish platform libraries

ibex088 committed:
Merge branch 'main' into feat/api-v2-extend-event-types-advanced-tab

SomayChauhan commented:
✅ **Published 0.0.38**

ibex088 committed:
Update documentation.json

supalarry committed:
fix: reset platform libraries to 0.0.0

ThyMinimalDev committed:
fixup! fix: reset platform libraries to 0.0.0

ibex088 committed:
Update event-type.tranformed.ts
